### ISSUE TITLE
torch.cuda.amp.GradScaler not working with ROCm — is there an alternative?

### ISSUE THREAD

**Issue by grptrzs** [ROCm/ROCm#6351]
Trying to use PyTorch AMP with ROCm but `GradScaler` seems to produce NaN losses after a few iterations. Same training loop works fine on NVIDIA.

Is AMP fully supported on ROCm? Or should I use a different approach for mixed precision on AMD GPUs?

```python
from torch.cuda.amp import autocast, GradScaler
scaler = GradScaler()  # This seems to cause issues

with autocast():
    loss = model(x)
scaler.scale(loss).backward()
scaler.step(optimizer)
scaler.update()
```

GPU: RX 7900 XTX (gfx1100)
ROCm: 5.7
PyTorch: 2.1.0+rocm5.7

**Comment by adityas-amd** [comment]
I reproduced on gfx1100 (PyTorch 2.7.1 / ROCm 7.x), in fp16 vs bf16 vs fp32, I found:
Normally-scaled model: fp16+GradScaler, bf16, and fp32 all stay finite and the loss decreases and AMP works.
When forward activations exceed 65504 (float16's max finite value): fp16 gives NaN within a few iters, while bf16 and fp32 stay finite (range ~3.4e38). Only the dtype differs.

Float16 has a narrow range (max 65504). Any forward activation/loss that overflows it becomes inf to NaN. GradScaler only rescales gradients; it can't keep a forward value finite, so it can't prevent this.
I would suggest to prefer bfloat16, drop GradScaler entirely and bf16 has fp32's exponent range and is native on RDNA3, so it doesn't overflow and needs no scaler:

```
with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
    loss = model(x)
    loss.backward()
    optimizer.step()
    optimizer.zero_grad()
```
  If you must use fp16, keep GradScaler but use the current API (torch.cuda.amp.* is deprecated) and tame the forward range (normalization, lower LR, grad clipping, watch exp/softmax/large logits):

``` 
 scaler = torch.amp.GradScaler("cuda")
  with torch.autocast(device_type="cuda", dtype=torch.float16):
      loss = model(x)
  scaler.scale(loss).backward()
  scaler.step(optimizer)
  scaler.update()
```

The fp16-overflow root cause is pure IEEE float math, so the bf16 fix works on your ROCm 5.7 too. That said, 5.7 is fairly old and RDNA3 support was early there, so it's worth retrying on a recent ROCm/PyTorch. In my testing on a current stack, switching to dtype=torch.bfloat16 and removing GradScaler resolves it.
